cpp/include/raft/linalg/detail/gemm.hpp

```diff
@@ -13,231 +13,398 @@
+ *
+ * @param [in] res raft resources
  * @param [in] trans_a cublas transpose op for A
  * @param [in] trans_b cublas transpose op for B
  * @param [in] m number of rows of C
  * @param [in] n number of columns of C
  * @param [in] k number of rows of opB(B) / number of columns of opA(A)
  * @param [in] alpha host or device scalar
- * @param [in] A such a matrix that the shape of column-major opA(A) is [m, k]
+ * @param [in] a_ptr such a matrix that the shape of column-major opA(A) is [m, k]
  * @param [in] lda leading dimension of A
- * @param [in] B such a matrix that the shape of column-major opA(B) is [k, n]
+ * @param [in] b_ptr such a matrix that the shape of column-major opA(B) is [k, n]
  * @param [in] ldb leading dimension of B
  * @param [in] beta host or device scalar
- * @param [inout] C column-major matrix of size [m, n]
+ * @param [inout] c_ptr column-major matrix of size [m, n]
  * @param [in] ldc leading dimension of C
  * @param [in] stream
  */
-template <typename math_t, bool DevicePointerMode = false>
-void gemm(raft::resources const& handle,
-          const bool trans_a,
-          const bool trans_b,
-          const int m,
-          const int n,
-          const int k,
-          const math_t* alpha,
-          const math_t* A,
-          const int lda,
-          const math_t* B,
-          const int ldb,
-          const math_t* beta,
-          math_t* C,
-          const int ldc,
-          cudaStream_t stream)
+template <bool DevicePointerMode = false, typename S, typename A, typename B, typename C>
+void matmul(raft::resources const& res,
+            bool trans_a,
+            bool trans_b,
+            uint64_t m,
+            uint64_t n,
+            uint64_t k,
+            const S* alpha,
+            const A* a_ptr,
+            uint64_t lda,
+            const B* b_ptr,
+            uint64_t ldb,
+            const S* beta,
+            C* c_ptr,
+            uint64_t ldc,
+            cudaStream_t stream)
 {
-  auto cublas_h = raft::resource::get_cublas_handle(handle);
-  cublas_device_pointer_mode<DevicePointerMode> pmode(cublas_h);
-  RAFT_CUBLAS_TRY(cublasgemm(cublas_h,
-                             trans_a ? CUBLAS_OP_T : CUBLAS_OP_N,
-                             trans_b ? CUBLAS_OP_T : CUBLAS_OP_N,
-                             m,
-                             n,
-                             k,
-                             alpha,
-                             A,
-                             lda,
-                             B,
-                             ldb,
-                             beta,
-                             C,
-                             ldc,
-                             stream));
+  common::nvtx::range<common::nvtx::domain::raft> batch_scope(
+    "linalg::matmul(m = %d, n = %d, k = %d)", m, n, k);
+  std::shared_ptr<matmul_desc> mm_desc{nullptr};
+  matmul_key_t mm_key{m, n, k, lda, ldb, ldc, trans_a, trans_b};
+  static thread_local cache::
```

Comment:
cc @robertmaynard for thoughts on implications to static link library (and potential conflics between shared state across applications).

Reply:
The issue around caching is due to these line from cublas docs:
```
A cuBLAS library context is tightly coupled with the CUDA context that is current at the time of the cublasCreate() call. An application that uses multiple CUDA contexts is required to create a cuBLAS context per CUDA context and make sure the former never outlives the latter.
```

So in our case we need all functions that call cublas or cache cublas types to be hidden symbols so that we maintain separate versions in different DSO as those will have separate CUDA contexts. The current version would need to have everything marked with `visibility("hidden")`

Reply:
Regarding the visibility of the functions: this PR does not change anything in that regard, so it's orthogonal to any solution we come up with.

Regarding the visibility of the static/thread_local variable declared *inside the function*. Am I right that it's not visible by default, so no action is needed here? I assume if RAFT is statically/dynamically linked multiple times, the variables like this won't collide. Also, since it is a locally defined variable, it is instantiated only upon the first call to the function.

Reply:
Resolved: avoiding the issue by keeping the cache inside the raft resources handle.